simplify JALR/Sentry text for RV64 (#994)
riscv/riscv-cheri: src/cheri/riscv-unpriv-integration.adoc

In the future, 32-bit bases ({cheri_base32_ext_name}) will also be defined with near identically behavior but a different in-memory representation, so this chapter uses the term {cheri_base_ext_name} to refer to the behavior common between all bases.
The {cheri_base_ext_name} base can be composed with other standard options to bases such as E (16-registers), Zfinx and endianness.

All CHERI implementations must also specify a capability encoding format.

This chapter also specifies <<rv64y_cap_description>> which is the capability encoding format for {cheri_base64_ext_name} systems.

ifdef::cheri_ratification_v1_only[]

> Future versions of this specification will include capability encoding formats for {cheri_base32_ext_name}}

Reviewer: ```suggestion
Future versions of this specification will include capability encoding formats for {cheri_base32_ext_name}
```

How about just keeping this as a non-normative NOTE:?

* The _base_ is returned by the <<GCBASE>> instruction.
* The _length_ is returned by the <<GCLEN>> instruction.

> ** <<GCLEN>> saturates the _length_ to XLEN bits

Author: YTOPR also saturates (to be confirmed @PeterRugg ), so this comment can probably go - as it was related to how you calculate the top

* The _base_ is returned by the <<GCBASE>> instruction.
* The _length_ is returned by the <<GCLEN>> instruction.
** <<GCLEN>> saturates the _length_ to XLEN bits

> * The _top_ is returned by the <<GCTOP>> instruction.

Reviewer: ```suggestion
* The _top_ is returned by the <<GCTOP>> instruction.
** <<GCTOP>> saturates the _length_ to XLEN bits.
```

This metadata value indicates the type of the capability and determines which operations the capability authorizes.
Which capability types a given CHERI platform supports is a function of the extensions and <<sec_cap_encoding_formats>> in use.
The capability encoding specifies a mapping between some bits within the capability format (usually described as "the CT field") and the capability types.

The capability encoding must be able to encode type `0` (_unsealed_).

> At least one non-zero type must also be implemented for control-flow integrity.

Reviewer 2: Instead: "If other sealed type exists, then we can provide different levels of control-flow integrity. If a default sealed type exists, then the "SENTRY" instruction can seal automatically to that type".

Author: I added the first sentence but not the second, and I'ves topped declaring sentry type 1 in the table as CHERIoT uses that value differently

Instructions that operate on capabilities will produce a result with the {ctag} set to zero if the source capability is sealed and the operation would alter its address, bounds, or permissions.
Extensions that extend the capability metadata must describe their interaction(s) with sealed capabilities.

Given a capability with `CT=0`,
deriving a capability with `CT≠0` is termed _sealing_
(or _sealing with type ``x``_ when a particular output ``CT=x`` is meant).

In the other direction,
deriving a `CT=0` capability from a `CT≠0` capability is termed _unsealing_
(or _unsealing from type ``x``_ when a particular input ``CT=x`` is meant).

[#sentry_cap,reftext="sentry capability"]
Sentry capability type::
Sentry capabilities provide *immutable* function pointers within a CHERI software system.
<<sealed_cap,Sealed capabilities>> are a natural foundation, providing immutability.

NOTE: Capabilities sealed with a specific <<sec_cap_type>> which can be unsealed when passed as an argument to <<JALR_CHERI>> are dubbed "sentries" (a portmanteau of "sealed entries").

Sentries can establish a form of control-flow integrity between mutually distrusting code.

> #TODO: ARC COMMENT: "sentries" is a bad choice for a shortened form of "sealed entries" as the standard english or even CS connotation of sentry doesn't match with a sealed entry.#

Reviewer 2: I mean, they do "lookout" for bad jumps.

Author: true! we need to resolve this.

Author: let's go with sentry

NOTE: In addition to using sealed capabilities as sentries for secure entry points, sealed capabilities can also be useful to software as secure software tokens.
 <<YSUNSEAL>> can be used to convert such a token back to an unsealed capability.
 A future extension may add an unseal instruction for performance.

The standard capability type mapping is shown in <<table_capability_types>>.

[#table_capability_types]

> .Default capability type mapping

Reviewer 2: Replace "Default" with "Standard"

[width="80%",options=header,align=left,cols="1,1,5"]
|===
| Type Name      | Integer Value | Description
| Unsealed       | 0             | _Unsealed_ capability granting access to memory

> | <<sentry_cap>> | 1             | Immutable _unrestricted sentry_ type for both _forward-edge_ and _backward-edge_ transitions.

Reviewer 2: Add "Also serves as the default sealed type in "RV64CHERI" encoding". Certain encodings, like CHERIoT, will not have a default sealed type.

Author: I fixed that by removing the integer value from the table

|===

The <<sec_cap_type>> value of 1 is enabled by the <<section_rvy_sentry_insn_ext>> extension.

Any sentry type may be passed as the `{cs1}` argument to <<JALR_CHERI>>, and are _unsealed_ on deference.
The returned `{cd}` value is also _sealed_ as a sentry.

Sentry types may be restricted to _forward-edge_ only and _backward-edge_ only transitions to further improve control-flow integrity.

> * When <<JALR_CHERI>> is used as a function call only _forward-edge_ sentries in `{cs1}` are _unsealed_ on deference.

Reviewer 2: "dereference", not "deference"

 The returned `{cd}` value is always _sealed_ as a _backward-edge_ sentry type.
* When <<JALR_CHERI>> is used as a function return only _backward-edge_ sentries in `{cs1}` are _unsealed_ on dereference.

> NOTE: Only a single sentry type is defined in <<table_capability_types>>, which is used for both _forward-edge_ and _backward-edge_ transitions.

Reviewer 2: Add "And this is also the "default sealing type" for the "RV64" encoding.".

Author: I've removed that note

 It is made use of in <<section_rvy_sentry_insn_ext>>.
 Capability formats may allocate more <<sec_cap_type>> values which are designated as sentries, and may distinguish them between _forward-edge_ and _backward-edge_ types.

NOTE: The integer value in the table above defines the architectural capability type (i.e., the result of the <<GCTYPE>> instruction) but does not need to be encoded that way in memory.

ifndef::cheri_ratification_v1_only[]

[NOTE]
====
In general, each of sealing and unsealing actions may require *authority*
to operate on the non-zero type;

> capability encoding formats which define types other than the ones in <<table_capability_types>> will specify how software expresses this authority.

Reviewer 2: Instead, say "If default sealing type is not defined, then each sealing and unsealing actions may require authority to operate on the non-zero type ...".?

Author: ok - I've removed the "default sealing type" from the spec entirely and it all makes much more sense now

NOTE: In addition to using sealed capabilities as sentries for secure entry points, sealed capabilities can also be useful to software as secure software tokens.
 <<YSUNSEAL>> can be used to convert such a token back to an unsealed capability.
 A future extension may add an unseal instruction for performance.

[#sec_cap_type_ambient, reftext="Ambient sealing type"]
Ambient sealing type::
Some capability types are said to be _ambiently available_ (or just _ambient_)
if they do not require specific authority to seal a capability with that type.
Therefore sentry types can be declared _ambient_.
For example, if <<section_rvy_sentry_insn_ext>> is available on a given platform,
the type with which it seals capabilities is considered _ambiently available_.

The standard capability type mapping is shown in <<table_capability_types>>, extensions and capability encoding formats may add more types.

[#table_capability_types]
.Standard capability type mapping
[width="80%",options=header,align=left,cols="1,1,5"]
|===
| Type Name      | Integer Value        | Description
| Unsealed       | 0                    | _Unsealed_ capability granting access to memory

> | <<sentry_cap>> | extension defined^1^ | Immutable _ambient sentry_ type for both _forward-edge_ and _backward-edge_ transitions.

Reviewer 3: "<<sentry_cap>>" was meant to be a property of otypes, not a name of an otype per se?  In CHERIoT, for example, we use type names like "backwards-edge interrupt-enabling sentry".

> | <<sentry_cap>> | extension defined^1^ | Immutable _ambient sentry_ type for both _forward-edge_ and _backward-edge_ transitions.

Reviewer: ```suggestion
| Unrestricted <<sentry_cap>> | extension defined^1^ | Immutable _ambient sentry_ type for both _forward-edge_ and _backward-edge_ transitions.
```
Maybe use this term?

> | <<sentry_cap>> | extension defined    | Immutable _sentry_ type for _forward-edge_ transitions.

Author: ```suggestion
| forward-edge <<sentry_cap>> | extension defined    | Immutable _sentry_ type for _forward-edge_ transitions.
```

> | <<sentry_cap>> | extension defined    | Immutable _sentry_ type for _backward-edge_ transitions.

Author: ```suggestion
| backward edge <<sentry_cap>> | extension defined    | Immutable _sentry_ type for _backward-edge_ transitions.
```

> | backward edge <<sentry_cap>> | extension defined    | Immutable _sentry_ type for _backward-edge_ transitions.

Reviewer: ```suggestion
| Forward-edge <<sentry_cap>> | extension defined    | Immutable _sentry_ type for _forward-edge_ transitions.
| Backward edge <<sentry_cap>> | extension defined    | Immutable _sentry_ type for _backward-edge_ transitions.
```

 A future extension may add an unseal instruction for performance.

[#sec_cap_type_ambient, reftext="Ambient sealing type"]
Ambient sealing type::
Some capability types are said to be _ambiently available_ (or just _ambient_)
if they do not require specific authority to seal a capability with that type.
Therefore sentry types can be declared _ambient_.
For example, if <<section_rvy_sentry_insn_ext>> is available on a given platform,
the type with which it seals capabilities is considered _ambiently available_.

The standard capability type mapping is shown in <<table_capability_types>>, extensions and capability encoding formats may add more types.

[#table_capability_types]
.Standard capability type mapping
[width="100%",options=header,align=left,cols="2,1,4"]
|===
| Type Name                     | Integer Value | Description
| Unsealed                      | 0             | _Unsealed_ capability granting access to memory
| Unrestricted <<sentry_cap>>^1^| 1             | Immutable _unrestricted sentry_ type for both _forward-edge_ and _backward-edge_ transitions.
| Forward-edge  <<sentry_cap>>  | 1             | Immutable _sentry_ type for _forward-edge_ transitions.

> | Backward-edge <<sentry_cap>>  | 1             | Immutable _sentry_ type for _backward-edge_ transitions.

Reviewer: ```suggestion
| Forward-edge  <<sentry_cap>>  | Defined by future extension | Immutable _sentry_ type for _forward-edge_ transitions.
| Backward-edge <<sentry_cap>>  | Defined by future extension | Immutable _sentry_ type for _backward-edge_ transitions.
```
We don't have those two types yet? They only exist in CHERIoT so far.

Reviewer: Or maybe just N/A and another footnote underneath to say

^2^ These types are not available in the current set of extensions but are available on CHERIoT systems.

Author: true - and value `1` isn't allocated here - I'll fix that now....

Sealed capabilities::
Capabilities with `CT≠0` are sealed against modification and cannot be dereferenced to access memory.
Instructions that operate on capabilities will produce a result with a cleared {ctag} if the source capability is sealed and the operation would alter its address, bounds, or permissions.
Extensions that extend the capability metadata must describe their interaction(s) with sealed capabilities.

Given a capability with `CT=0`,
deriving a capability with `CT≠0` is termed "sealing"
(or "sealing with type ``x``" when a particular output ``CT=x`` is meant).

In the other direction,
deriving a `CT=0` capability from a `CT≠0` capability is termed "unsealing"
(or "unsealing from type ``x``" when a particular input ``CT=x`` is meant).

ifndef::cheri_ratification_v1_only[]

[NOTE]
====
In general, each of sealing and unsealing actions may require *authority*
to operate on the non-zero type;

> capability encoding formats which define types other than in <<table_capability_types>> will specify how software expresses this authority.

Reviewer: ```suggestion
capability encoding formats which define types other than the ones in <<table_capability_types>> will specify how software expresses this authority.
```

Sealed capabilities::
Capabilities with `CT≠0` are sealed against modification and cannot be dereferenced to access memory.
Instructions that operate on capabilities will produce a result with a cleared {ctag} if the source capability is sealed and the operation would alter its address, bounds, or permissions.

> Extensions that manipulate capability metadata must describe their interaction(s) with sealed capabilities.

Reviewer: ```suggestion
Extensions that extend the capability metadata must describe their interaction(s) with sealed capabilities.
```

the <<x_perm>> in the capability encoding effectively adds
an additional bit to the <<sec_cap_type>> field.

Capability encoding formats will explicitly state whether these are relevant to them.

They are not used by <<rv64y_cap_description>>, but are by the CHERIoT formats.
====

[#sec_cap_type_ambient]
Ambient sealing type::
Some capability types are said to be "ambiently available" (or just "ambient")
if they do not require specific authority to seal a capability (with that type).

> For example, if the <<YSENTRY>> instruction is available on a given platform,

Reviewer: ```suggestion
For example, if the <<SENTRY>> instruction is available on a given platform,
```

Author: I missed this one - I've reverted to the original text.

the <<x_perm>> in the capability encoding effectively adds
an additional bit to the <<sec_cap_type>> field.

Capability encoding formats will explicitly state whether these are relevant to them.

They are not used by <<rv64y_cap_description>>, but are by the CHERIoT formats.
====

[#sec_cap_type_ambient]
Ambient sealing type::
Some capability types are said to be "ambiently available" (or just "ambient")
if they do not require specific authority to seal a capability (with that type).
For example, if <<section_rvy_sentry_insn_ext>> is available on a given platform,
the type with which it seals capabilities is considered ambiently available.
(For the capability encoding format specified by <<rv64y_cap_description>>, that would be type `1`.)

> endif::[]

Reviewer 2: Remove the word "ambiently available". Instead stick to if "default sealed type" is defined, then do this ...

Author: I've removed the default sealed type and so removed that statement

The {rvy_sentry_insn_ext_name} extension:

> . Defines one <<sentry_cap>> type, the <<sentry_cap>> type for both _forward-edge_ and _backward-edge_ transitions, with a <<sec_cap_type>> value of `1`.

Reviewer: ```suggestion
. Defines one <<sentry_cap>> type, the <<sentry_cap>> type for both _forward-edge_ and _backward-edge_ control-flow transitions, with a <<sec_cap_type>> value of `1`.
```

> . Defines one <<sentry_cap>> type, the <<sec_cap_type_ambient,ambient>> sentry type for both _forward-edge_ and _backward-edge_ transitions, with a <<sec_cap_type>> value of `1`.

Reviewer 3: ```suggestion
. Defines one <<sentry_cap>> type, the <<sec_cap_type_ambient,ambient>> <<sentry_cap>> type for both _forward-edge_ and _backward-edge_ transitions, with a <<sec_cap_type>> value of `1`.
```

> . Adds the <<SENTRY>> instruction to allow <<sec_cap_type_ambient,ambient>> sealing of capabilities as sentries with <<sec_cap_type>> value of `1`.

Reviewer: ```suggestion
. Adds the <<SENTRY>> instruction to allow <<sec_cap_type_ambient,ambient sealing >> of capabilities as sentries with <<sec_cap_type>> value of `1`.
```

{rvy_sentry_insn_ext_name} is only compatible with capability encoding formats which can encode the <<sec_cap_type>> value of `1`.

=== Interaction with <<JALR_CHERI>>

{rvy_sentry_insn_ext_name} enables _sealing_ and _unsealing_ behavior in <<JALR_CHERI>> with the single <<sentry_cap>> of type `1` being available.

Therefore the full operational description of <<JALR_CHERI>> is specialized to:

. `{cs1}` is written to the target <<pcc>>.
. The target address is obtained by adding the sign-extended 12-bit I-immediate to `{cs1}.address`, then setting the least-significant bit of the result to zero.
. _Unseal_ the target <<pcc>> if it is _sealed_ as a _forward-edge_ <<sentry_cap>> (type `1` qualifies for this) and `{cs1}.address[0]` is zero and the I-immediate is zero.
. Set the address of the target <<pcc>> to the target address using the semantics of the <<SCADDR>> instruction.
. The <<pcc>> of the next instruction is _sealed_ as a <<sentry_cap>> of type `1` and written to `{cd}`.
. Jump to the target <<pcc>>.

> NOTE: Future extensions may add other <<sentry_cap>> types, but to alter the behavior of <<JALR_CHERI>> differently to this then they will be incompatible with {rvy_sentry_insn_ext_name}.

Reviewer: I think this note is confusing, I'd just drop it.
```suggestion
```

=== Interaction with <<JAL_CHERI>>

{rvy_sentry_insn_ext_name} enables _sealing_ behavior in <<JAL_CHERI>> with the single <<sentry_cap>> of type `1` being available.

Therefore the sealed type of `{cd}` is defined as follows:

* The <<pcc>> of the next instruction is _sealed_ as a <<sentry_cap>> of type `1` and written to `{cd}`.

NOTE: Future extensions may add other <<sentry_cap>> types, but if they alter the sealed type written by <<JAL_CHERI>> then they will be incompatible with {rvy_sentry_insn_ext_name}.

> 

Reviewer: ```suggestion
```

Again I would drop this to avoid confusion for readers, and just note in the Zycheriot section that it is incompatible (since JALR seals differently).

=== Interaction with <<JALR_CHERI>>

> {rvy_sentry_insn_ext_name} adds sealing and unsealing behavior to <<JALR_CHERI>>.

Reviewer: I'm not sure we can do this while removing the mention of hooks in the jalr. My understanding is that unprivileged extensions can't redefine behaviour of existing instructions.

Author: I think this is something to discuss with ARC - _how_ to describe this stuff - I'll give it another go first

ifdef::cheri_ratification_v1_only[]

> * If `{cs1}` is _sealed_ then _unseal_ it before installing in <<pcc>> (see <<sec_cap_type>>).

Reviewer: And target address== cs1.addr

Author: as a note - as adding the address increment clears the tag as part of YADDRW - so this is already covered

* The <<pcc>> of the next instruction is _sealed_ as a <<sentry_cap>> then then written to `{cd}`.
endif::[]

ifndef::cheri_ratification_v1_only[]
. For HOOK 2: Seal the return capability as a sentry.
endif::[]

NOTE: When any sealed capability is passed as the input to <<JALR_CHERI>>, its address must have bit zero clear and the instruction must have a zero offset, or the target <<pcc>> will have its {ctag} set to zero, since updates to its address are interpreted with the semantics of the <<SCADDR>> instruction.
 As such, a <<sentry_cap>> defines a secure function entry point, and as such the offset in the JALR instruction must be zero.

> NOTE: Future {cheri_base_ext_name} extensions are permitted to introduce additional restrictions on the behaviour of <<JALR_CHERI>>, by extending the number of available values in the <<sec_cap_type>> field in the capability encoding format.

Reviewer: Why is this note in zysentry?
